File: stripe_terminal/ios/Classes/StripeTerminalPlugin.swift
import Flutter
import StripeTerminal
import UIKit

public class StripeTerminalPlugin: NSObject, FlutterPlugin, StripeTerminalPlatformApi {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = StripeTerminalPlugin(registrar.messenger())
        setStripeTerminalPlatformApiHandler(registrar.messenger(), instance)
        instance.onAttachedToEngine()
    }
    
    private let handlers: StripeTerminalHandlersApi
    
    init(_ binaryMessenger: FlutterBinaryMessenger) {
        self.handlers = StripeTerminalHandlersApi(binaryMessenger)
        self._discoverReadersController = DiscoverReadersControllerApi(binaryMessenger: binaryMessenger)
        self._readerDelegate = ReaderDelegatePlugin(handlers)
        self._readerReconnectionDelegate = ReaderReconnectionDelegatePlugin(handlers)
    }

    public func onAttachedToEngine() {
        self.setupDiscoverReaders()
    }
    
    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        self._discoverReadersController.removeHandler()
        removeStripeTerminalPlatformApiHandler()
        self._clean()
    }
    
    func onInit(_ shouldPrintLogs: Bool) async throws {
        // If a hot restart is performed in flutter the terminal is already initialized but we need to clean it up
        if Terminal.hasTokenProvider() {
            _clean()
            return
        }
        
        let delegate = TerminalDelegatePlugin(handlers)
        Terminal.setTokenProvider(delegate)
        Terminal.shared.delegate = delegate
        if (shouldPrintLogs) { Terminal.setLogListener { message in print(message) } }
    }
    
    func onClearCachedCredentials() throws {
        Terminal.shared.clearCachedCredentials()
    }

// Reader discovery, connection and updates
    private let _discoverReadersController: DiscoverReadersControllerApi
    private let _discoveryDelegate = DiscoveryDelegatePlugin()
    private var _readers: [Reader] { get {
        return _discoveryDelegate.readers
    } }
    private let _readerDelegate: ReaderDelegatePlugin
    private let _readerReconnectionDelegate: ReaderReconnectionDelegatePlugin

    func onGetConnectionStatus() throws -> ConnectionStatusApi {
        return Terminal.shared.connectionStatus.toApi()
    }
    
    func onSupportsReadersOfType(
        _ deviceType: DeviceTypeApi,
        _ discoveryMethod: DiscoveryMethodApi,
        _ simulated: Bool
    ) throws -> Bool {
        let hostDiscoveryMethod = discoveryMethod.toHost()
        guard let hostDiscoveryMethod else {
            return false
        }
        let hostDeviceType = deviceType.toHost()
        guard let hostDeviceType else {
            return false
        }
        let result = Terminal.shared.supportsReaders(
            of: hostDeviceType,
            discoveryMethod: hostDiscoveryMethod,
            simulated: simulated
        )
        do {
            try result.get()
            return true
        } catch {
            return false
        }
    }
    
    func setupDiscoverReaders() {
        _discoverReadersController.setHandler(
            _discoveryDelegate.onListen,
            _discoveryDelegate.onCancel
        )
    }

    func onConnectHandoffReader(_ serialNumber: String) async throws -> ReaderApi {
        throw PlatformError("", "Unsupported method")
    }

    func onConnectBluetoothReader(
        _ serialNumber: String,
        _ locationId: String,
        _ autoReconnectOnUnexpectedDisconnect: Bool
    ) async throws -> ReaderApi {
        do {
            let reader = try await Terminal.shared.connectBluetoothReader(
                _findReader(serialNumber),
                delegate: _readerDelegate,
                connectionConfig: BluetoothConnectionConfiguration(
                    locationId: locationId,
                    autoReconnectOnUnexpectedDisconnect: autoReconnectOnUnexpectedDisconnect,
                    autoReconnectionDelegate: _readerReconnectionDelegate
                )
            )
            return reader.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }

    func onConnectInternetReader(
        _ serialNumber: String,
        _ failIfInUse: Bool
    ) async throws -> ReaderApi {
        do {
            let reader = try await Terminal.shared.connectInternetReader(
                _findReader(serialNumber),
                connectionConfig: InternetConnectionConfiguration(
                    failIfInUse: failIfInUse
                )
            )
            return reader.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }

    func onConnectMobileReader(
        _ serialNumber: String,
        _ locationId: String
    ) async throws -> ReaderApi {
        do {
            let reader = try await Terminal.shared.connectLocalMobileReader(
                _findReader(serialNumber),
                delegate: _readerDelegate,
                connectionConfig: LocalMobileConnectionConfiguration(
                    locationId: locationId
                )
            )
            return reader.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }
    
    func onConnectUsbReader(_ serialNumber: String, _ locationId: String, _ autoReconnectOnUnexpectedDisconnect: Bool) async throws -> ReaderApi {
        throw PlatformError("", "Unsupported method")
    }

    func onGetConnectedReader() throws -> ReaderApi? {
        return Terminal.shared.connectedReader?.toApi()
    }
    
    func onCancelReaderReconnection() async throws {
        try await _readerReconnectionDelegate.cancelable?.cancel()
    }
    
    func onListLocations(_ endingBefore: String?, _ limit: Int?, _ startingAfter: String?) async throws -> [LocationApi] {
        do {
            return try await Terminal.shared.listLocations(parameters: ListLocationsParameters(
                limit: limit as NSNumber?,
                endingBefore: endingBefore,
                startingAfter: startingAfter
            )).0.map { $0.toApi() }
        } catch let error as NSError {
            throw error.toApi()
        }
    }
    
    func onInstallAvailableUpdate() throws {
        Terminal.shared.installAvailableUpdate()
    }
    
    func onCancelReaderUpdate() async throws {
        try await _readerDelegate.cancellableUpdate?.cancel()
    }
    
    func onDisconnectReader() async throws {
        do {
            try await Terminal.shared.disconnectReader()
        } catch let error as NSError {
            throw error.toApi()
        }
    }
    
// Taking payments
    
    private var _paymentIntents: [String: PaymentIntent] = [:]
    
    func onGetPaymentStatus() throws -> PaymentStatusApi {
        return Terminal.shared.paymentStatus.toApi()
    }
    
    func onCreatePaymentIntent(_ parameters: PaymentIntentParametersApi) async throws -> PaymentIntentApi {
        do {
            let paymentIntent = try await Terminal.shared.createPaymentIntent(parameters.toHost())
            _paymentIntents[paymentIntent.stripeId] = paymentIntent
            return paymentIntent.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }

    func onRetrievePaymentIntent(
        _ clientSecret: String
    ) async throws -> PaymentIntentApi {
        do {
            let paymentIntent = try await Terminal.shared.retrievePaymentIntent(clientSecret: clientSecret)
            _paymentIntents[paymentIntent.stripeId] = paymentIntent
            return paymentIntent.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }

    private var _cancelablesCollectPaymentMethod: [Int: Cancelable] = [:]

    func onStartCollectPaymentMethod(
        _ result: Result<PaymentIntentApi>,
        _ operationId: Int,
        _ paymentIntentId: String,
        _: Bool,
        _: Bool
    ) throws {
        let paymentIntent = try _findPaymentIntent(paymentIntentId)
        self._cancelablesCollectPaymentMethod[operationId] = Terminal.shared.collectPaymentMethod(paymentIntent) { paymentIntent, error in
            self._cancelablesCollectPaymentMethod.removeValue(forKey: operationId)
            if let error = error as? NSError {
                let platformError = error.toApi()
                result.error(platformError.code, platformError.message, platformError.details)
                return
            }
            self._paymentIntents[paymentIntent!.stripeId] = paymentIntent!
            result.success(paymentIntent!.toApi())
        }
    }

    func onStopCollectPaymentMethod(
        _ operationId: Int
    ) async throws {
        try await _cancelablesCollectPaymentMethod.removeValue(forKey: operationId)?.cancel()
    }

    func onProcessPayment(
        _ paymentIntentId: String
    ) async throws -> PaymentIntentApi {
        let paymentIntent = try _findPaymentIntent(paymentIntentId)
        do {
            let (intent, error) = await Terminal.shared.processPayment(paymentIntent)
            if let error {
                throw PlatformError(error.declineCode!, error.localizedDescription)
            }
            return intent!.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }
    
    func onCancelPaymentIntent(_ paymentIntentId: String) async throws -> PaymentIntentApi {
        do {
            let paymentIntent = try _findPaymentIntent(paymentIntentId)
            let newPaymentIntent = try await Terminal.shared.cancelPaymentIntent(paymentIntent)
            _paymentIntents.removeValue(forKey: paymentIntentId)
            return newPaymentIntent.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }
    
// Saving payment details for later use
   
    private var _cancelablesReadReusableCard: [Int: Cancelable] = [:]
    private var _setupIntents: [String: SetupIntent] = [:]
    private var _cancelablesCollectSetupIntentPaymentMethod: [Int: Cancelable] = [:]

    func onStartReadReusableCard(
        _ result: Result<PaymentMethodApi>,
        _ operationId: Int,
        _: String?,
        _: [String: String]?
    ) throws {
        _cancelablesReadReusableCard[operationId] = Terminal.shared.readReusableCard(
            ReadReusableCardParameters(),
            completion: { paymentMethod, error in
                self._cancelablesReadReusableCard.removeValue(forKey: operationId)
                if let error = error as? NSError {
                    let platformError = error.toApi()
                    result.error(platformError.code, platformError.message, platformError.details)
                    return
                }
                result.success(paymentMethod!.toApi())
            }
        )
    }

    func onStopReadReusableCard(
        _ operationId: Int
    ) async throws {
        try await _cancelablesReadReusableCard.removeValue(forKey: operationId)?.cancel()
    }
    
    func onCreateSetupIntent(
        _ customerId: String?,
        _ metadata: [String : String]?,
        _ onBehalfOf: String?,
        _ description: String?,
        _ usage: SetupIntentUsageApi?
    ) async throws -> SetupIntentApi {
        let params = SetupIntentParameters(customer: customerId)
        params.metadata = metadata
        params.onBehalfOf = onBehalfOf
        params.stripeDescription = description
        if let usage = usage { params.usage = usage.toHost() }
        do {
            let setupIntent = try await Terminal.shared.createSetupIntent(params)
            _setupIntents[setupIntent.stripeId] = setupIntent
            return setupIntent.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }
    
    func onRetrieveSetupIntent(_ clientSecret: String) async throws -> SetupIntentApi {
        do {
            let setupIntent = try await Terminal.shared.retrieveSetupIntent(clientSecret: clientSecret)
            _setupIntents[setupIntent.stripeId] = setupIntent
            return setupIntent.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }
    
    func onStartCollectSetupIntentPaymentMethod(
        _ result: Result<SetupIntentApi>,
        _ operationId: Int,
        _ setupIntentId: String,
        _ customerConsentCollected: Bool
    ) throws {
        let setupIntent = try _findSetupIntent(setupIntentId)
        _cancelablesCollectSetupIntentPaymentMethod[operationId] = Terminal.shared.collectSetupIntentPaymentMethod(
            setupIntent,
            customerConsentCollected: customerConsentCollected,
            completion: { setupIntent, error in
                self._cancelablesCollectSetupIntentPaymentMethod.removeValue(forKey: operationId)
                if let error = error as? NSError {
                    let platformError = error.toApi()
                    result.error(platformError.code, platformError.message, platformError.details)
                    return
                }
                self._setupIntents[setupIntent!.stripeId] = setupIntent!
                result.success(setupIntent!.toApi())
            }
        )
    }
    
    func onStopCollectSetupIntentPaymentMethod(_ operationId: Int) async throws {
        try await _cancelablesCollectSetupIntentPaymentMethod.removeValue(forKey: operationId)?.cancel()
    }
    
    func onConfirmSetupIntent(_ setupIntentId: String) async throws -> SetupIntentApi {
        let setupIntent = try _findSetupIntent(setupIntentId)
        let (newSetupIntent, error) = await Terminal.shared.confirmSetupIntent(setupIntent)
        if let error {
            throw PlatformError(error.declineCode!, error.localizedDescription)
        }
        _setupIntents[newSetupIntent!.stripeId] = newSetupIntent!
        return newSetupIntent!.toApi()
    }
    
    func onCancelSetupIntent(_ setupIntentId: String) async throws -> SetupIntentApi {
        let setupIntent = try _findSetupIntent(setupIntentId)
        do {
            let newSetupIntent = try await Terminal.shared.cancelSetupIntent(setupIntent)
            _setupIntents.removeValue(forKey: setupIntentId)
            return newSetupIntent.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }
// Card-present refunds
    private var _cancelablesCollectRefundPaymentMethod: [Int: Cancelable] = [:]

    func onStartCollectRefundPaymentMethod(
        _ result: Result<Void>,
        _ operationId: Int,
        _ chargeId: String,
        _ amount: Int,
        _ currency: String,
        _ metadata: [String : String]?,
        _ reverseTransfer: Bool?,
        _ refundApplicationFee: Bool?
    ) throws {
        let params = RefundParameters(chargeId: chargeId, amount: UInt(amount), currency: currency)
        if let metadata = metadata { params.metadata = metadata }
        if let reverseTransfer = reverseTransfer { params.reverseTransfer = NSNumber(value: reverseTransfer) }
        if let refundApplicationFee = refundApplicationFee { params.refundApplicationFee = NSNumber(value: refundApplicationFee) }
        _cancelablesCollectRefundPaymentMethod[operationId] = Terminal.shared.collectRefundPaymentMethod(
            params,
            completion: { error in
                self._cancelablesCollectRefundPaymentMethod.removeValue(forKey: operationId)
                if let error = error as? NSError {
                    let platformError = error.toApi()
                    result.error(platformError.code, platformError.message, platformError.details)
                    return
                }
                result.success(())
            }
        )
    }
    
    func onStopCollectRefundPaymentMethod(_ operationId: Int) async throws {
        try await _cancelablesCollectRefundPaymentMethod.removeValue(forKey: operationId)?.cancel()
    }
    
    func onProcessRefund() async throws -> RefundApi {
        do {
            let (refund, error) = await Terminal.shared.processRefund()
            if let error {
                throw PlatformError("\(error.code)", error.localizedDescription)
            }
            return refund!.toApi()
        } catch let error as NSError {
            throw error.toApi()
        }
    }

// Display information to customers
    
    func onSetReaderDisplay(
        _ cart: CartApi
    ) async throws {
        do {
            try await Terminal.shared.setReaderDisplay(cart.toHost())
        } catch let error as NSError {
            throw error.toApi()
        }
    }

    func onClearReaderDisplay() async throws {
        do {
            try await Terminal.shared.clearReaderDisplay()
        } catch let error as NSError {
            throw error.toApi()
        }
    }
    
    // PRIVATE METHODS
    
    private func _clean() {
        if (Terminal.shared.connectedReader != nil) { Terminal.shared.disconnectReader  { error in } }
        
        self._discoveryDelegate.clear()
        
        self._cancelablesCollectPaymentMethod.values.forEach { $0.cancel { error in } }
        self._cancelablesCollectPaymentMethod = [:]
        self._paymentIntents = [:]
        
        self._cancelablesReadReusableCard.values.forEach { $0.cancel { error in } }
        self._cancelablesReadReusableCard = [:]
        self._cancelablesCollectSetupIntentPaymentMethod.values.forEach { $0.cancel { error in } }
        self._cancelablesCollectSetupIntentPaymentMethod = [:]
        self._setupIntents = [:]

        self._cancelablesCollectRefundPaymentMethod.values.forEach { $0.cancel { error in } }
        self._cancelablesCollectRefundPaymentMethod = [:]
    }

    private func _findReader(_ serialNumber: String) throws -> Reader {
        guard let reader = _readers.first(where: { $0.serialNumber == serialNumber }) else {
            throw PlatformError(
                TerminalExceptionCodeApi.readerCommunicationError.rawValue,
                "Reader with provided serial number no longer exists"
            )
        }
        return reader
    }
    
    private func _findPaymentIntent(_ paymentIntentId: String) throws -> PaymentIntent {
        let paymentIntent = _paymentIntents[paymentIntentId]
        guard let paymentIntent else {
            throw PlatformError(TerminalExceptionCodeApi.paymentIntentNotRetrieved.rawValue, nil, nil)
        }
        return paymentIntent
    }
    
    private func _findSetupIntent(_ setupIntentId: String) throws -> SetupIntent {
        let setupIntent = _setupIntents[setupIntentId]
        guard let setupIntent else {
            throw PlatformError(TerminalExceptionCodeApi.paymentIntentNotRetrieved.rawValue, nil, nil)
        }
        return setupIntent
    }
}

File: stripe_terminal/android/src/main/kotlin/com/stripe_terminal/StripeTerminalPlugin.kt
package com.stripe_terminal

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.ContextCompat
import com.stripe.stripeterminal.Terminal
import com.stripe.stripeterminal.TerminalApplicationDelegate
import com.stripe.stripeterminal.external.callable.Callback
import com.stripe.stripeterminal.external.callable.Cancelable
import com.stripe.stripeterminal.external.callable.ConnectionTokenCallback
import com.stripe.stripeterminal.external.callable.ConnectionTokenProvider
import com.stripe.stripeterminal.external.callable.DiscoveryListener
import com.stripe.stripeterminal.external.callable.LocationListCallback
import com.stripe.stripeterminal.external.callable.PaymentIntentCallback
import com.stripe.stripeterminal.external.callable.PaymentMethodCallback
import com.stripe.stripeterminal.external.callable.ReaderCallback
import com.stripe.stripeterminal.external.callable.TerminalListener
import com.stripe.stripeterminal.external.models.CollectConfiguration
import com.stripe.stripeterminal.external.models.ConnectionConfiguration
import com.stripe.stripeterminal.external.models.ConnectionStatus
import com.stripe.stripeterminal.external.models.ConnectionTokenException
import com.stripe.stripeterminal.external.models.DeviceType
import com.stripe.stripeterminal.external.models.DiscoveryConfiguration
import com.stripe.stripeterminal.external.models.DiscoveryMethod
import com.stripe.stripeterminal.external.models.ListLocationsParameters
import com.stripe.stripeterminal.external.models.Location
import com.stripe.stripeterminal.external.models.PaymentIntent
import com.stripe.stripeterminal.external.models.PaymentMethod
import com.stripe.stripeterminal.external.models.PaymentStatus
import com.stripe.stripeterminal.external.models.ReadReusableCardParameters
import com.stripe.stripeterminal.external.models.Reader
import com.stripe.stripeterminal.external.models.TerminalException
import com.stripe.stripeterminal.log.LogLevel
import com.stripe_terminal.api.CartApi
import com.stripe_terminal.api.ConnectionStatusApi
import com.stripe_terminal.api.DiscoverReadersControllerApi
import com.stripe_terminal.api.DiscoveryMethodApi
import com.stripe_terminal.api.LocationApi
import com.stripe_terminal.api.PaymentIntentApi
import com.stripe_terminal.api.PaymentMethodApi
import com.stripe_terminal.api.PlatformException
import com.stripe_terminal.api.ReaderApi
import com.stripe_terminal.api.Result
import com.stripe_terminal.api.StripeTerminalHandlersApi
import com.stripe_terminal.api.StripeTerminalPlatformApi
import com.stripe_terminal.api.toApi
import com.stripe_terminal.api.toHost
import com.stripe_terminal.plugin.ReaderDelegatePlugin
import com.stripe_terminal.plugin.ReaderReconnectionListenerPlugin
import com.stripe_terminal.plugin.TerminalErrorHandler
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger

class StripeTerminalPlugin : FlutterPlugin, ActivityAware,
    StripeTerminalPlatformApi,
    ConnectionTokenProvider, TerminalListener {
    private lateinit var _handlers: StripeTerminalHandlersApi
    private lateinit var _discoverReadersController: DiscoverReadersControllerApi

    private val _terminal: Terminal get() = Terminal.getInstance()

    private var _activity: Activity? = null
    private var _discoveredReaders: List<Reader> = arrayListOf()

    private val permissions = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
        arrayOf(
            Manifest.permission.BLUETOOTH_SCAN,
            Manifest.permission.BLUETOOTH_ADMIN,
        )
    } else {
        arrayOf(
            Manifest.permission.BLUETOOTH_ADMIN,
        )
    }

    override fun onInit() {
        val permissionStatus = permissions.map {
            ContextCompat.checkSelfPermission(_activity!!, it)
        }

        if (permissionStatus.contains(PackageManager.PERMISSION_DENIED)) {
            throw PlatformException(
                "stripeTerminal#permissionDeclinedPermanenty",
                "You have declined the necessary permission, please allow from settings to continue.",
                null
            )
        }

        // If a hot restart is performed in flutter the terminal is already initialized but we need to clean it up
        if (Terminal.isInitialized()) {
            clean()
            return
        }

        Terminal.initTerminal(
            _activity!!.applicationContext,
            LogLevel.NONE,
            this,
            this,
        )
    }

    //region Terminal listeners
    override fun fetchConnectionToken(callback: ConnectionTokenCallback) {
        _activity!!.runOnUiThread {
            _handlers.requestConnectionToken({ code, message, details ->
                val exception = PlatformException(code, message, details)
                callback.onFailure(ConnectionTokenException("", exception))
            }, { token ->
                callback.onSuccess(token)
            })
        }
    }
    //endregion

    //region Reader discovery, connection and updates
    private var _discoverReaderCancelable: Cancelable? = null
    private val _readerDelegate = ReaderDelegatePlugin(_handlers)
    private val _readerReconnectionDelegate = ReaderReconnectionListenerPlugin(_handlers)

    override fun onConnectionStatusChange(status: ConnectionStatus) {
        _activity!!.runOnUiThread {
            _handlers.connectionStatusChange(status.toApi())
        }
    }

    override fun onConnectionStatus(): ConnectionStatusApi {
        return _terminal.connectionStatus.toApi()
    }

    override fun onUnexpectedReaderDisconnect(reader: Reader) {
        _activity!!.runOnUiThread {
            _handlers.unexpectedReaderDisconnect(reader.toApi())
        }
    }

    private fun setupDiscoverReadersController(binaryMessenger: BinaryMessenger) {
        _discoverReadersController = DiscoverReadersControllerApi(binaryMessenger)
        _discoverReadersController.setHandler({ sink, discoveryMethod: DiscoveryMethodApi, simulated: Boolean, locationId: String? ->
            val hostDiscoveryMethod = discoveryMethod.toHost()
            if (hostDiscoveryMethod == null) {
                sink.error("discoveryMethodNotSupported", null, null)
                sink.endOfStream()
                return@setHandler
            }

            val config = DiscoveryConfiguration(
                isSimulated = simulated,
                discoveryMethod = hostDiscoveryMethod,
                location = locationId
            )

            // Ignore error, the previous stream can no longer receive events
            _discoverReaderCancelable?.cancel(EmptyCallback())

            _discoverReaderCancelable =
                _terminal.discoverReaders(config, object : DiscoveryListener {
                    override fun onUpdateDiscoveredReaders(readers: List<Reader>) {
                        _discoveredReaders = readers
                        _activity!!.runOnUiThread { sink.success(readers.map { it.toApi() }) }
                    }
                }, object : TerminalErrorHandler(sink::error), Callback {
                    override fun onFailure(e: TerminalException) = _activity!!.runOnUiThread {
                        super.onFailure(e)
                        sink.endOfStream()
                    }

                    override fun onSuccess() = _activity!!.runOnUiThread { sink.endOfStream() }
                })
        }, {
            // Ignore error, flutter stream already closed
            _discoverReaderCancelable?.cancel(EmptyCallback())
        })
    }

    override fun onConnectBluetoothReader(
        result: Result<ReaderApi>,
        serialNumber: String,
        locationId: String,
        autoReconnectOnUnexpectedDisconnect: Boolean,
    ) {
        val reader = findActiveReader(result, serialNumber) ?: return

        _terminal.connectBluetoothReader(reader,
            ConnectionConfiguration.BluetoothConnectionConfiguration(
                locationId = locationId,
                autoReconnectOnUnexpectedDisconnect = autoReconnectOnUnexpectedDisconnect,
                bluetoothReaderReconnectionListener = _readerReconnectionDelegate,
            ),
            _readerDelegate,
            object : TerminalErrorHandler(result::error), ReaderCallback {
                override fun onSuccess(reader: Reader) = result.success(reader.toApi())
            }
        )
    }

    override fun onConnectHandoffReader(
        result: Result<ReaderApi>,
        serialNumber: String,
    ) {
        val reader = findActiveReader(result, serialNumber) ?: return

        _terminal.connectHandoffReader(reader,
            ConnectionConfiguration.HandoffConnectionConfiguration(),
            _readerDelegate,
            object : TerminalErrorHandler(result::error), ReaderCallback {
                override fun onSuccess(reader: Reader) = result.success(reader.toApi())
            }
        )
    }

    override fun onConnectInternetReader(
        result: Result<ReaderApi>,
        serialNumber: String,
        failIfInUse: Boolean
    ) {
        val reader = findActiveReader(result, serialNumber) ?: return

        _terminal.connectInternetReader(reader,
            ConnectionConfiguration.InternetConnectionConfiguration(
                failIfInUse = failIfInUse,
            ),
            object : TerminalErrorHandler(result::error), ReaderCallback {
                override fun onSuccess(reader: Reader) = result.success(reader.toApi())
            })
    }

    override fun onConnectMobileReader(
        result: Result<ReaderApi>,
        serialNumber: String,
        locationId: String,
    ) {
        val reader = findActiveReader(result, serialNumber) ?: return

        val config = ConnectionConfiguration.LocalMobileConnectionConfiguration(
            locationId = locationId,
        )
        _terminal.connectLocalMobileReader(reader, config,
            object : TerminalErrorHandler(result::error), ReaderCallback {
                override fun onSuccess(reader: Reader) = result.success(reader.toApi())
            })
    }

    override fun onConnectUsbReader(
        result: Result<ReaderApi>,
        serialNumber: String,
        locationId: String,
        autoReconnectOnUnexpectedDisconnect: Boolean,
    ) {
        val reader = findActiveReader(result, serialNumber) ?: return

        _terminal.connectUsbReader(reader,
            ConnectionConfiguration.UsbConnectionConfiguration(
                locationId = locationId,
                autoReconnectOnUnexpectedDisconnect = autoReconnectOnUnexpectedDisconnect,
                usbReaderReconnectionListener = _readerReconnectionDelegate,
            ),
            _readerDelegate,
            object : TerminalErrorHandler(result::error), ReaderCallback {
                override fun onSuccess(reader: Reader) = result.success(reader.toApi())
            }
        )
    }

    override fun onConnectedReader(): ReaderApi? {
        return _terminal.connectedReader?.toApi()
    }

    override fun onCancelReaderReconnection(result: Result<Unit>) {
        if (_readerReconnectionDelegate.cancelReconnect == null) {
            result.success(Unit)
        }
        _readerReconnectionDelegate.cancelReconnect?.cancel(object : Callback, TerminalErrorHandler(result::error) {
            override fun onSuccess() = result.success(Unit)
        })
    }

    override fun onListLocations(
        result: Result<List<LocationApi>>,
        endingBefore: String?,
        limit: Long?,
        startingAfter: String?,
    ) {
        val params = ListLocationsParameters.Builder()
        params.endingBefore = endingBefore
        params.startingAfter = startingAfter
        params.limit = limit?.toInt()
        _terminal.listLocations(params.build(),
            object : TerminalErrorHandler(result::error), LocationListCallback {
                override fun onSuccess(locations: List<Location>, hasMore: Boolean) =
                    result.success(locations.map { it.toApi() })
            })
    }

    override fun onInstallAvailableUpdate() {
        _terminal.installAvailableUpdate()
    }

    override fun onCancelReaderUpdate(result: Result<Unit>) {
        if (_readerDelegate.cancelUpdate == null) {
            result.success(Unit)
        }
        _readerDelegate.cancelUpdate?.cancel(object : Callback, TerminalErrorHandler(result::error) {
            override fun onSuccess() = result.success(Unit)
        })
    }

    override fun onDisconnectReader(result: Result<Unit>) {
        _terminal.disconnectReader(object : TerminalErrorHandler(result::error), Callback {
            override fun onSuccess() = result.success(Unit)
        })
    }
    //endregion

    //region Taking Payment
    private var _paymentIntents = HashMap<String, PaymentIntent>()

    override fun onPaymentStatusChange(status: PaymentStatus) {
        _activity!!.runOnUiThread {
            _handlers.paymentStatusChange(status.toApi())
        }
    }

    override fun onRetrievePaymentIntent(
        result: Result<PaymentIntentApi>,
        clientSecret: String
    ) {
        _terminal.retrievePaymentIntent(clientSecret,
            object : TerminalErrorHandler(result::error), PaymentIntentCallback {
                override fun onSuccess(paymentIntent: PaymentIntent) {
                    _paymentIntents[paymentIntent.id] = paymentIntent
                    result.success(paymentIntent.toApi())
                }
            })
    }

    private val _cancelablesCollectPaymentMethod = HashMap<Long, Cancelable>()

    override fun onStartCollectPaymentMethod(
        result: Result<PaymentIntentApi>,
        operationId: Long,
        paymentIntentId: String,
        moto: Boolean,
        skipTipping: Boolean,
    ) {
        val paymentIntent = _paymentIntents[paymentIntentId]
        if (paymentIntent == null) {
            result.error("", "")
            return
        }
        _cancelablesCollectPaymentMethod[operationId] = _terminal.collectPaymentMethod(
            paymentIntent,
            object : TerminalErrorHandler(result::error), PaymentIntentCallback {
                override fun onFailure(e: TerminalException) {
                    _cancelablesCollectPaymentMethod.remove(operationId)
                    super.onFailure(e)
                }

                override fun onSuccess(paymentIntent: PaymentIntent) {
                    _cancelablesCollectPaymentMethod.remove(operationId)
                    result.success(paymentIntent.toApi())
                    _paymentIntents[paymentIntent.id] = paymentIntent
                }
            },
            CollectConfiguration.Builder()
                .setMoto(moto)
                .skipTipping(skipTipping)
                .build(),
        )
    }

    override fun onStopCollectPaymentMethod(result: Result<Unit>, operationId: Long) {
        _cancelablesCollectPaymentMethod.remove(operationId)
            ?.cancel(object : TerminalErrorHandler(result::error), Callback {
                override fun onSuccess() = result.success(Unit)
            })
    }

    override fun onProcessPayment(
        result: Result<PaymentIntentApi>,
        paymentIntentId: String,
    ) {
        val paymentIntent = _paymentIntents[paymentIntentId]
        if (paymentIntent == null) {
            result.error("", "")
            return
        }
        _terminal.processPayment(
            paymentIntent,
            object : TerminalErrorHandler(result::error), PaymentIntentCallback {
                override fun onSuccess(paymentIntent: PaymentIntent) {
                    result.success(paymentIntent.toApi())
                    _paymentIntents.remove(paymentIntent.id)
                }
            })
    }
    //endregion

    //region Saving payment details for later use
    private val _cancelablesReadReusableCard = HashMap<Long, Cancelable>()

    override fun onStartReadReusableCard(
        result: Result<PaymentMethodApi>,
        operationId: Long,
        customer: String?,
        metadata: HashMap<String, String>?,
    ) {
        val params = ReadReusableCardParameters.Builder()
        if (customer != null) params.setCustomer(customer)
        if (metadata != null) params.putAllMetadata(metadata)
        _cancelablesReadReusableCard[operationId] = _terminal.readReusableCard(params.build(),
            object : TerminalErrorHandler(result::error), PaymentMethodCallback {
                override fun onFailure(e: TerminalException) {
                    _cancelablesReadReusableCard.remove(operationId)
                    super.onFailure(e)
                }

                override fun onSuccess(paymentMethod: PaymentMethod) {
                    _cancelablesReadReusableCard.remove(operationId)
                    result.success(paymentMethod.toApi())
                }
            })
    }

    override fun onStopReadReusableCard(
        result: Result<Unit>,
        operationId: Long,
    ) {
        _cancelablesReadReusableCard.remove(operationId)
            ?.cancel(object : TerminalErrorHandler(result::error), Callback {
                override fun onSuccess() = result.success(Unit)
            })
    }
    //endregion

    //region Display information to customers
    override fun onSetReaderDisplay(result: Result<Unit>, cart: CartApi) {
        _terminal.setReaderDisplay(cart.toHost(),
            object : TerminalErrorHandler(result::error), Callback {
                override fun onSuccess() = result.success(Unit)
            })
    }

    override fun onClearReaderDisplay(result: Result<Unit>) {
        _terminal.clearReaderDisplay(object : TerminalErrorHandler(result::error), Callback {
            override fun onSuccess() = result.success(Unit)
        })
    }
    //endregion

    //region Android
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val binaryMessenger = flutterPluginBinding.binaryMessenger
        StripeTerminalPlatformApi.setHandler(binaryMessenger, this)
        _handlers = StripeTerminalHandlersApi(binaryMessenger)

        setupDiscoverReadersController(binaryMessenger)
    }

    override fun onDetachedFromEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        clean()

        _discoverReadersController.removeHandler()
        StripeTerminalPlatformApi.removeHandler()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        _activity = binding.activity
        TerminalApplicationDelegate.onCreate(_activity!!.application)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        _activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        _activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        _activity = null
    }
    //endregion

    // ======================== INTERNAL METHODS

    private fun findActiveReader(result: Result<*>, serialNumber: String): Reader? {
        val reader = _discoveredReaders.firstOrNull { it.serialNumber == serialNumber }
        if (reader == null) {
            result.error(
                TerminalException.TerminalErrorCode.READER_CONNECTED_TO_ANOTHER_DEVICE.name,
                "Reader with provided serial number no longer exists",
                null
            )
        }
        return reader
    }

    private fun clean() {
        if (_terminal.connectedReader != null) _terminal.disconnectReader(EmptyCallback())

        _discoverReaderCancelable?.cancel(EmptyCallback())
        _discoverReaderCancelable = null

        _cancelablesReadReusableCard.values.forEach { it.cancel(EmptyCallback()) }
        _cancelablesReadReusableCard.clear()
        _cancelablesCollectPaymentMethod.values.forEach { it.cancel(EmptyCallback()) }
        _cancelablesCollectPaymentMethod.clear()

        _discoveredReaders = arrayListOf()
        _paymentIntents = hashMapOf()
    }
}

class EmptyCallback : Callback {
    override fun onFailure(e: TerminalException) {}
    override fun onSuccess() {}
}
